applications/nrf_desktop/src/modules/hid_forward.c

```diff
@@ -199,66 +156,120 @@ static int register_subscriber(struct bt_gatt_dm *dm, uint16_t pid,
+		 * on error.
+		 */
+		uint16_t recipient = sub->cfg_chan_rsp->recipient;
+		uint8_t event_id = sub->cfg_chan_rsp->event_id;
+
+		int pos = config_channel_report_parse(data,
+						      REPORT_SIZE_USER_CONFIG,
+						      sub->cfg_chan_rsp);
 
-		int pos = config_channel_report_parse(data, REPORT_SIZE_USER_CONFIG,
-						      &frame, false);
 		if (pos < 0) {
-			LOG_WRN("Could not set report");
+			LOG_WRN("Failed to parse response: %d", pos);
+			sub->cfg_chan_rsp->recipient = recipient;
+			sub->cfg_chan_rsp->event_id = event_id;
+			submit_forward_error_rsp(sub, CONFIG_STATUS_WRITE_ERROR);
 			return pos;
 		}
 
-		if (frame.status != CONFIG_STATUS_SUCCESS) {
-			LOG_INF("GATT read done, but fetch was not ready yet");
+		if (sub->cfg_chan_rsp->status == CONFIG_STATUS_PENDING) {
+			LOG_WRN("GATT read done, but fetch was not ready yet");
+			sub->cfg_chan_rsp->recipient = recipient;
+			sub->cfg_chan_rsp->event_id = event_id;
+			sub->cur_poll_cnt++;
 
-			/* Do not notify requester, host will schedule next read. */
-			forward_pending = false;
-			return 0;
+			if (sub->cur_poll_cnt >= CFG_CHAN_MAX_RSP_POLL_CNT) {
+				submit_forward_error_rsp(sub, CONFIG_STATUS_WRITE_ERROR);
+			} else {
+				k_delayed_work_submit(&sub->read_rsp,
+						      CFG_CHAN_RSP_READ_DELAY);
+			}
+		} else {
+			EVENT_SUBMIT(sub->cfg_chan_rsp);
+			sub->cfg_chan_rsp = NULL;
 		}
+	}
 
-		struct config_fetch_event *event =
-			new_config_fetch_event(frame.event_data_len);
+	return 0;
+}
 
-		event->id = frame.event_id;
-		event->recipient = frame.recipient;
-		event->channel_id = channel_id;
+static void read_rsp_fn(struct k_work *work)
+{
+	struct hids_subscriber *sub = CONTAINER_OF(work,
+						struct hids_subscriber,
+						read_rsp);
+	struct bt_gatt_hids_c_rep_info *config_rep =
+		bt_gatt_hids_c_rep_find(&sub->hidc,
+					BT_GATT_HIDS_REPORT_TYPE_FEATURE,
+					REPORT_ID_USER_CONFIG);
 
-		memcpy(event->dyndata.data, &data[pos], frame.event_data_len);
+	__ASSERT_NO_MSG(config_rep);
+	int err = bt_gatt_hids_c_rep_read(&sub->hidc,
+					  config_rep,
+					  hidc_read_cfg);
 
-		EVENT_SUBMIT(event);
+	if (err) {
+		LOG_WRN("Cannot read feature report (err: %d)", err);
+		submit_forward_error_rsp(sub, CONFIG_STATUS_WRITE_ERROR);
 	}
+}
 
-	return 0;
+static void hidc_write_cb(struct bt_gatt_hids_c *hidc,
+			  struct bt_gatt_hids_c_rep_info *rep,
+			  uint8_t err)
+{
+	struct hids_subscriber *sub = CONTAINER_OF(hidc,
+						   struct hids_subscriber,
+						   hidc);
+
+	if (err) {
+		LOG_WRN("Failed to write report: %d", err);
+		submit_forward_error_rsp(sub, CONFIG_STATUS_WRITE_ERROR);
+		return;
+	}
+
+	/* HID forward does not fetch configuration channel response for
+	 * set operation. This is done to prevent HID report rate drops
+	 * (for LLPM connection, the peripheral may send either HID report
+	 * or configuration channel response in given connection interval).
+	 */
+	if (sub->cfg_chan_rsp->status != CONFIG_STATUS_SET) {
```

Comment:
I'm not sure if we need a separate variable to handle operation status here. It can simply be kept as the response status. The response status is updated only right before sending the response.